src/lib.rs

```diff
@@ -891,290 +869,176 @@ impl<L: SynthLanguage> Synthesizer<L> {
         }
     }
 
-    // Run single chunk
-    fn run_chunk_rule_lifting(&mut self, chunk: &[RecExpr<L>]) {
-        Synthesizer::add_chunk(&mut self.egraph, chunk);
-
-        let mut new_allowed_eqs: EqualityMap<L> = EqualityMap::default();
-        let mut new_forbidden_eqs: EqualityMap<L> = EqualityMap::default();
-        for (name, eq) in self.new_eqs.clone() {
-            if L::is_allowed_rewrite(&eq.lhs, &eq.rhs) {
-                new_allowed_eqs.insert(name, eq);
-            } else {
-                new_forbidden_eqs.insert(name, eq);
+    fn run_rule_lifting(mut self) -> Report<L> {
+        let timer = Instant::now();
+        if let Some(filename) = &self.params.workload {
+            let terms = self.enumerate_workload(filename);
+            for chunk in terms.chunks(self.params.node_chunk_size) {
+                self.run_chunk_rule_lifting(chunk);
             }
+        } else {
+            let iters = self
+                .params
+                .iters
+                .expect("Either iters or workload is required");
+            for iter in 1..=iters {
+                let layer = self.enumerate_layer(iter);
+                for chunk in layer.chunks(self.params.node_chunk_size) {
+                    self.run_chunk_rule_lifting(chunk);
+                }
+            }
+        }
+
+        let time = timer.elapsed().as_secs_f64();
+
+        let mut all_eqs: Vec<Equality<L>> =
+            self.all_eqs.clone().into_iter().map(|(_, eq)| eq).collect();
+        all_eqs.sort_by_key(|eq| eq.score());
+        all_eqs.reverse();
+        let mut new_eqs: Vec<Equality<L>> =
+            self.new_eqs.clone().into_iter().map(|(_, eq)| eq).collect();
+        new_eqs.sort_by_key(|eq| eq.score());
+        new_eqs.reverse();
+        let mut old_eqs: Vec<Equality<L>> =
+            self.old_eqs.clone().into_iter().map(|(_, eq)| eq).collect();
+        old_eqs.sort_by_key(|eq| eq.score());
+        old_eqs.reverse();
+
+        for eq in &new_eqs {
+            println!("  {:?}   {}", eq.score(), eq);
         }
+        println!(
+            "Learned {} rules in {:?} using {} old rules.",
+            new_eqs.len(),
+            time,
+            old_eqs.len()
+        );
+
+        Report {
+            params: self.params,
+            time,
+            num_rules: new_eqs.len(),
+            all_eqs,
+            new_eqs,
+            old_eqs,
+            smt_unknown: self.smt_unknown,
+        }
+    }
 
-        let mut old_allowed_eqs: EqualityMap<L> = EqualityMap::default();
-        let mut old_forbidden_eqs: EqualityMap<L> = EqualityMap::default();
-        for (name, eq) in self.old_eqs.clone() {
+    fn run_chunk_rule_lifting(&mut self, chunk: &[RecExpr<L>]) {
+        log::info!("Adding {} terms to egraph", chunk.len());
```

Comment:
is this `step 1. Add workload terms to egraph ` from the ruler doc? If so, can we add a comment?